tee_supplicant/src/tee_supplicant.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/tee.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <optee_msg_supplicant.h>
+#include <tee_client_api.h>
+#include <teec_ta_load.h>
+
+LOG_MODULE_REGISTER(tee_supplicant);
+
+#define TEE_SUPP_THREAD_PRIO	7
+
+#define TEE_REQ_PARAM_MAX	5
+
+static struct k_thread main_thread;
+static K_THREAD_STACK_DEFINE(main_stack, 8192);
+
+static K_MUTEX_DEFINE(shm_mutex);
+
+#define MEMREF_SHM_ID(p)	((p)->c)
+#define MEMREF_SHM_OFFS(p)	((p)->a)
+#define MEMREF_SIZE(p)		((p)->b)
+
+struct tee_supp_msg {
+	uint32_t cmd_ret;
+	uint32_t num_param;
+	struct tee_param params[TEE_REQ_PARAM_MAX];
+};
+
+struct param_value {
+	uint64_t a;
+	uint64_t b;
+	uint64_t c;
+};
+
+static int receive_request(const struct device *dev, struct tee_supp_msg *ts_req)
+{
+	int rc = tee_suppl_recv(dev, &ts_req->cmd_ret, &ts_req->num_param, ts_req->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant receive failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static int send_response(const struct device *dev, struct tee_supp_msg *rsp)
+{
+	int rc = tee_suppl_send(dev, rsp->cmd_ret, rsp->num_param, rsp->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant send response failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static void uuid_from_octets(TEEC_UUID *d, const uint8_t s[TEE_UUID_LEN])
+{
+	d->timeLow = (s[0] << 24) | (s[1] << 16) | (s[2] << 8) | s[3];
+	d->timeMid = (s[4] << 8) | s[5];
+	d->timeHiAndVersion = (s[6] << 8) | s[7];
+	memcpy(d->clockSeqAndNode, s + 8, sizeof(d->clockSeqAndNode));
+}
+
+static int load_ta(uint32_t num_params, struct tee_param *params)
+{
+	int ta_found = 0;
+	size_t size = 0;
+	uint8_t val_cmd[TEE_UUID_LEN];
+	TEEC_UUID uuid = { 0 };
+	struct tee_shm *shm;
+	void *addr;
+
+	if (num_params != 2) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	if ((params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) != TEE_PARAM_ATTR_TYPE_VALUE_INPUT ||
+	    (params[1].attr & TEE_PARAM_ATTR_TYPE_MASK) != TEE_PARAM_ATTR_TYPE_MEMREF_OUTPUT) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	memcpy(val_cmd, &params[0].a, TEE_UUID_LEN);
+	shm = (struct tee_shm *)params[1].c;
+
+	uuid_from_octets(&uuid, val_cmd);
```

Comment:
```suggestion
	shm = (struct tee_shm *)params[1].c;

	uuid_from_param(&uuid, params[0].a, params[0].b);
```
And then in uuid_from_param params[0].b is copied to d->clockSeqAndNode and params[0].a splits to u32 and 2 u16 as mentioned in RFC4122. This will overcome padding issue. Also, as I saw - this problem is also present on the original optee_client. I think an issue should be created or so...

Reply:
Ok